File: web-app/src/types/metricAnomalies.ts
/**
 * Anomaly Inbox types — mirror the Rust `entity::metric_anomalies::Model`
 * and the `metric_anomalies` HTTP responses.
 */

export type AnomalySeverity = "low" | "medium" | "high";
export type AnomalyStatus = "new" | "acknowledged" | "dismissed";

/** A single dimension filter identifying which segment an anomaly belongs to. */
export interface AnomalyFilter {
  /** Fully-qualified dimension id, e.g. `"labor_daily.restaurant_id"`. */
  member: string;
  /** Matched values (OR within a filter). */
  values: string[];
}

export interface MetricAnomaly {
  id: string;
  workspace_id: string;
  measure: string;
  time_dimension: string;
  granularity: string;
  period_start: string;
  period_end: string;
  observed: number;
  expected: number;
  lower_bound: number;
  upper_bound: number;
  z_score: number;
  severity: AnomalySeverity;
  status: AnomalyStatus;
  label: string | null;
  /**
   * Stable key derived from the monitor's filters (e.g.
   * `"labor_daily.restaurant_id=loc-abc"`). Empty string for chain-wide
   * (unfiltered) monitors. Distinguishes per-segment anomalies that share a
   * measure/period so they don't read as duplicates.
   */
  dimension_key: string;
  /** Raw filters identifying this anomaly's segment. Null for chain-wide monitors. */
  filters: AnomalyFilter[] | null;
  /**
   * Groups consecutive flagged buckets of one segment into a single event, so a
   * surge spanning Mon/Wed/Thu reads as one problem rather than three. Rows stay
   * per-bucket (explain reasons about a single bucket), so the collapsing
   * happens here on read. Null for rows detected before events existed.
   */
  event_id: string | null;
  detected_at: string;
  updated_at: string;
}

export interface ListAnomaliesResponse {
  anomalies: MetricAnomaly[];
}

/** A monitor that errored during a scan — identifies the monitor/segment and the error. */
export interface ScanFailure {
  measure: string;
  time_dimension: string;
  granularity: string;
  label: string | null;
  /** Segment key when the failed monitor was a group_by/filtered segment; empty for chain-wide. */
  dimension_key: string;
  /** Raw filters identifying the segment; null for chain-wide monitors. */
  filters: AnomalyFilter[] | null;
  error: string;
}

export interface ScanAnomaliesResponse {
  monitors_scanned: number;
  monitors_failed: number;
  anomalies_persisted: number;
  /** True when the scan is still running in the background. Refetch anomalies after a short delay. */
  pending?: boolean;
  /** Per-monitor failures. Empty on a clean scan and on the `pending` path. */
  failures?: ScanFailure[];
}

export interface MonitorEntry {
  measure: string;
  time_dimension: string;
  granularity: "day" | "week" | "month";
  lookback_days: number;
  seasonality: number[] | null;
  sensitivity: "low" | "medium" | "high";
  label?: string | null;
  /**
   * Dimension filters narrowing this entry to one segment. Two entries over the
   * same measure/time-dimension/granularity are distinguished only by these, so
   * anything mapping coverage rows back to an entry must use them.
   */
  filters?: AnomalyFilter[];
  /**
   * Fan-out dimension: the scanner discovers its values at scan time and files
   * one coverage row per segment, each keyed by this entry's `filters` *plus*
   * the discovered value.
   */
  group_by?: string | null;
}

/** Per-segment scan coverage. A `group_by` monitor fans out to one row per
 *  segment, so a single `MonitorEntry` can map to many of these.
 *
 *  Exists because a monitor skipped for want of history produces neither an
 *  anomaly nor a failure — without this the UI cannot tell "healthy, nothing
 *  found" from "not scoring at all". */
export interface MonitorCoverage {
  id: string;
  workspace_id: string;
  measure: string;
  time_dimension: string;
  granularity: string;
  /** Empty string for chain-wide monitors. */
  dimension_key: string;
  filters: Record<string, unknown>[] | null;
  label: string | null;
  /** Buckets the warehouse returned; zero-filled gaps are not counted. */
  measured_buckets: number;
  /** The statistical floor this segment must clear to be scored. */
  required_buckets: number;
  last_scanned_at: string;
}

export interface ListMonitorsResponse {
  monitors: MonitorEntry[];
  coverage: MonitorCoverage[];
}

File: crates/entity/src/metric_anomalies.rs
//! Anomalies surfaced by `oxy-metric-monitoring`.
//!
//! One row per (workspace, measure, time-dim, period_start). Repeat scans
//! upsert via the unique index so an unresolved anomaly stays visible
//! without piling up duplicates. Status drives the inbox UX:
//! `new` (visible) → `acknowledged` (still visible, marked seen) →
//! `dismissed` (hidden).

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "metric_anomalies")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: Uuid,
    pub workspace_id: Uuid,
    pub measure: String,
    pub time_dimension: String,
    pub granularity: String,
    pub period_start: DateTimeWithTimeZone,
    pub period_end: DateTimeWithTimeZone,
    pub observed: f64,
    pub expected: f64,
    pub lower_bound: f64,
    pub upper_bound: f64,
    pub z_score: f64,
    pub severity: String,
    pub status: String,
    pub label: Option<String>,
    /// Stable key derived from the monitor's filters (e.g. `"sales_daily.restaurant_id=loc-abc"`).
    /// Empty string for chain-wide (unfiltered) monitors. Included in the
    /// unique index so per-store and chain-wide anomalies don't collide.
    pub dimension_key: String,
    /// The raw filters used when fetching the series, stored as JSON so the
    /// API can expose which store/segment this anomaly belongs to.
    #[sea_orm(column_type = "JsonBinary", nullable)]
    pub filters: Option<serde_json::Value>,
    /// Cached `airlayer::engine::metric_tree_ops::ExplainResult` for the
    /// drawer. `None` = never computed (or invalidated). Populated lazily
    /// by `POST /semantic/anomalies/{id}/explain`.
    #[sea_orm(column_type = "JsonBinary", nullable)]
    pub explain_cache: Option<serde_json::Value>,
    /// Wall-clock when [`Self::explain_cache`] was written. Lets callers
    /// reason about freshness without inspecting the cached payload.
    pub explain_cached_at: Option<DateTimeWithTimeZone>,
    /// Groups consecutive flagged buckets of one segment into a single event,
    /// so a labour surge spanning Mon/Wed/Thu reads as one problem rather than
    /// three. Assigned at persist time by proximity + direction; `None` for
    /// rows detected before events existed.
    ///
    /// Rows stay per-bucket on purpose — `explain_anomaly` reasons about a
    /// single bucket against the same phase one cycle back, so merging rows
    /// would make it describe only the first day of a range.
    pub event_id: Option<Uuid>,
    /// Groups anomalies that fired across *different* segments on the same
    /// bucket in the same direction, so a chain-wide collapse reads as one
    /// event rather than one row per store. Orthogonal to [`Self::event_id`],
    /// which chains consecutive buckets within a single segment — a multi-day
    /// chain-wide slide is both, which is why this is a separate column.
    ///
    /// Membership is recomputed from the scan that observed it rather than
    /// preserved (it is a property of that scan's share, not a historical fact),
    /// but the *identity* is deterministic: derived via `Uuid::new_v5` over the
    /// cohort key, so a restatement re-scan of the same bucket keeps the same id
    /// instead of minting a fresh one each day.
    pub cohort_id: Option<Uuid>,
    /// This member's ratio-to-expectation divided by the cohort's median ratio.
    /// 1.0 is a typical member. Direction-relative: for a *drop* cohort the
    /// actionable rows are **below** 1.0 (fell further than the shared event
    /// explains); for an *increase* cohort the outlier is the one **above** 1.0.
    /// A consumer ranking members must read the cohort's direction to know which
    /// tail to sort toward. `None` outside a cohort, or where no finite ratio
    /// exists.
    pub cohort_deviation: Option<f64>,
    /// The tenant calendar's name for the day this cohort landed on, e.g.
    /// `"Independence Day"`. A label the inbox shows, never a filter that
    /// suppresses the row — see the `m20260804_000001` migration.
    ///
    /// Denormalised at scan time rather than resolved on read: the calendar is
    /// editable config, and this should keep saying what the scan believed.
    pub cohort_label: Option<String>,
    /// Dominant seasonal cycle length (in units of [`Self::granularity`]) from
    /// the monitor's detection config, snapshotted at scan time. Drives the
    /// same-phase comparison window when explaining this anomaly. `None` for
    /// rows detected before this column existed.
    pub seasonal_period: Option<i32>,
    pub detected_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}
